### ISSUE TITLE
Load files from mainPackage first, then from the package list

### ISSUE THREAD

**Issue by strongholdmedia** [composer/composer#7535]
For what it's worth, it would seem totally logical that classes are loaded in a `packageList`-then-`mainPackage` order so that any classes provided by the vendor classes.
_However, this makes no difference, as with autoloading, a single class namespace should only be provided by a single file, and loading is on-demand, not as with e.g. C headers where the order does matter._

With classless files, though, this is totally different.

Since they can not really override functionality (as in PHP, neither functions nor constants may be overloaded), it is totally expectable that, in order for the `mainPackage` things take precedence over the `packageList` (vendor) ones, that `mainPackage` files be included *first*.

_I am not up to debating whether using global functions and variables at all is a good idea or not; I am fairly certain that the whole idea is horrific; but this is out of scope._

This being said, it would be greatly useful to have "files" loaded with `mainPackage`-then-`packageList` order.

*Of course*, I am aware that this *could* wreak havoc in any packages relying on this "undocumented feature" (i.e. the current behavior). But I am unsure to what extent.
If wide enough, this may mandate a configurable option.

I will send a quick PR, please comment with your thoughts.

(NB this is the exact same issue as #6768 - even though the wording is not clear there, nor are possibly the implications. I just opened a new issue as somebody else was requested to do.)

**Comment by strongholdmedia** [comment]
I added pull request https://github.com/composer/composer/pull/7536 but I think that there is even a test bound to current behavior.

**Comment by alcohol** [comment]
TL/DR; if the explicit ordering is put behind a flag/argument/config value, this could be implemented without breaking any potential BC (one could argue there is no BC though, but that is quite a topic for debate by itself).

---

I personally also believe that your proposal makes more sense, and I am quite certain that the current behaviour is not intentional, but merely "as it has always been". Just like in mysql --where it is often cautioned that not explicitly defining an `order by` clause means the order of the resultset can be entirely random and unpredictable and should not be relied upon-- I think the same applies here.

But then on the other hand, whether the current behaviour is explicit or not, one could also argue that the behaviour you propose is exactly that, explicit, and thus could just as easily be put behind a toggle/flag to enable or disable it.

The latter would also allow it to be implemented without impacting anything or anyone (by defaulting to the non-explicit ordering).

**Comment by alcohol** [comment]
@Seldaek do you have any thoughts on whether the current (non-explicit) behaviour should be maintained, or if we should make the explicit-order the default behaviour?

**Comment by strongholdmedia** [comment]
Totally agreed. I also endorse the configuration option unless it would be utterly useless.

**Comment by jvasseur** [comment]
I think the current order is the one that makes more sense, you might want to use some features from dependencies in your main package file. For example adding something in a global registry.

**Comment by alcohol** [comment]
> I think the current order is the one that makes more sense

That is entirely subjective, but I can see why, because

> you might want to use some features from dependencies in your main package file

is a very valid point.

So making this configurable and explicit is probably the best way to go, and the default should be as it is now.

**Comment by Seldaek** [comment]
The current order isn't random, it's defined by https://github.com/composer/composer/blob/7072b25c63652d04c5bda1a8c9bfe675ed7031c7/src/Composer/Autoload/AutoloadGenerator.php#L956 - and aims to load packages that are depended upon before those that are depending on them. 

If you need to define some things in your project before the things you depend on get a chance to load.. might I suggest you avoid abusing the autoloader and instead simply include a file doing the globa declares you want *before* you include the autoloader and load the rest? That would avoid us having to add a config flag for some edge case.

**Comment by strongholdmedia** [comment]
> might I suggest you avoid abusing the autoloader

Of course you might. _Even so, I'd argue that allowing to include random files without classes constitutes, in itself, "abuse of the autoloader"._
But, sadly, "big projects", like Laravel, started to abusing this feature in the exact same way I suggested here (as well as many others, as previous closed issues signify).

*And as you surely know, the open/closed principle suffers.*

Laravel is aimed for beginner-friendliness, who often have a hard time getting a grasp why someone should not do things (e.g. copying all the classes into a single file instead using an autoloader, or even changing files in libraries then leaving when upgrade comes, etc, etc).

Of course, I could just include a file here, then there, and leave 300 lines of comments everywhere that of course, THIS file is included from here, THEN composer loads random arbitrary files that can change overnight from random dependences that may change overnight, THEN I include finally THIS file, also manually, from a totally different place, that must absolutely be included last.

Or I could find another approach (which I did, as solutions are often needed fast).
Or I could just use reflection to find whether a global method has a given signature or not.
_Or I could even use runkit to unload a method and then declare it again._

Any of that would, however, be missing the point _entirely_.

(Not to mention that if I include a file before the autoloader, that wishes to also have PSR-x classes, then that would need a separate autoloader...)

**Comment by Seldaek** [comment]
Can you show one concrete example of what you need this for though? I am curious because maybe  I live in a different world here but I can only imagine a very few legit cases and they wouldn't lead to the chaos you describe IMO. 

The solution otherwise might be to educate those projects you are using and hope we all move to better practices instead of working around and piling on the tech debt.

**Comment by strongholdmedia** [comment]
I totally agree with you about educating those projects..

This specific issue is described in detail at the other issue: #6768

Surely enough, I exaggerated a bit the part on chaos, but still, the thing is that if I start including random files in the bootstrap file, then I have to answer **both** _why_ not to include _other_ files in the bootstrapper before the autoloader, _and_ _why_ are Laravel's files not included in their autoloader but with Composer - from every time thereon.

As I've mentioned earlier, I think this whole global helper function  (/ "facade") mayhem is an anti-pattern, but it seems to me it is unrealistic to expect Laravel usage to decrease.

**Comment by Seldaek** [comment]
Laravel probably should define a way to hook into it instead of overriding outright, because that leads to code duplication. There could be a global way to add hooks to the global helpers, in which case you could register your hook during app init and not have to fiddle with autoloading order.

Looking at the issue you linked though, I don't really see the problem with the additional require statement. Sure it breaks in phpunit if you don't define a custom bootstrap file.. Composer [also has a special bootstrap file](https://github.com/composer/composer/blob/master/tests/bootstrap.php) for phpunit, and it doesn't cause us any hassle.

I think this is overblown really, and IMO changing the order will cause more problems than anything. Of course you may disagree, but I am going to close this as a won't fix.

**Comment by chriscadcw** [comment]
I know this is a "closed" issue, but I think that this issue was closed before the discussion was complete.  I am all for better educating developers in best practices instead of convuluting the language.  That being said however, I can provide a perfect example of where this type of scenario, loading the "files" section from the project composer.json before loading the vendor package "files" sections would be beneficial.

@Seldaek you asked for a concrete example of where this would be required, here is what I am currently working with and stuck on.

We currently have a package in that is being required `"akaunting/money": "^1.0"` for our project.  In this package, their composer.json file includes a "files" section for their own helper functions:

```
"autoload":{  
      "psr-4":{  
         "Akaunting\\Money\\":"./src"
      },
      "files":[  
         "src/helpers.php"
      ]
   },
```

Within this helpers.php file are the following two function definitions:

```
use Akaunting\Money\Currency;
use Akaunting\Money\Money;

if (!function_exists('money')) {
    /**
     * Instance of money class.
     *
     * @param mixed  $amount
     * @param string $currency
     * @param bool   $convert
     *
     * @return \Akaunting\Money\Money
     */
    function money($amount, $currency = 'USD', $convert = false)
    {
        return new Money($amount, currency($currency), $convert);
    }
}

if (!function_exists('currency')) {
    /**
     * Instance of currency class.
     *
     * @param string $currency
     *
     * @return \Akaunting\Money\Currency
     */
    function currency($currency)
    {
        return new Currency($currency);
    }
}
```

These functions are designed in a way that a programmer should be able to override them, so in order to do so we originally had our own helper file with our custom definitions:

```
if(!function_exists('money')) {
    /**
     * Instance of money class.
     *
     * @param mixed $amount
     * @param string $currency
     * @param bool $convert
     *
     * @return \Akaunting\Money\Money
     */
    function money($amount, $currency = 'USD', $convert = false) {
        return new \App\Helpers\MoneyExtended($amount, currency($currency), $convert);
    }
}

if(!function_exists('currency')) {
    /**
     * Instance of currency class.
     *
     * @param string $currency
     *
     * @return \Akaunting\Money\Currency
     */
    function currency($currency) {
        return new \Akaunting\Money\Currency($currency);
    }
}
```

This allows us to specify that the `money()` function should use our class which extends the Akaunting\Money class whenever it is called.  To get around the issue that was presented in this thread to begin with, we simply updated our index.php file to read:

```

**Comment by greaterking** [comment]
@chriscadcw 

Additionally this would not only be true for tinker but also anything else outside the web routes  i.e the api/ routes as they do not use index.php. In this case the returned extended class defined in the money() function would return one set of data from the extended class to web routes and another set for the api routes (because the api route will use \Akaunting\Money and not \App\Helpers\MoneyExtended).
